README.md
# OftenOn PowerShell Module by Cody Konior

![OftenOn logo][1]

[![Build status](https://ci.appveyor.com/api/projects/status/smdxnxpi6c006son?svg=true)](https://ci.appveyor.com/project/codykonior/oftenon)

Read the [CHANGELOG][3]

## Description

OftenOn is a Lability/Hyper-V/Desired State Configuration wrapper which builds and configures five-node no-storage multi-subnet
Windows Server Failover Cluster(s) with SQL Server Availability Groups.

It was designed specifically to be compatible with Windows Server 2012 (WS2012) and SQL Server 2012 (SQL2012) and for testing
migration scenarios to higher clusters (a SQL2012 cross-cluster migration) and higher SQL Server versions (a SQL 2017 upgrade
followed by a Distributed Availability Group migration).

## Installation

- `Install-Module OftenOn`

## Major functions

- `Set-OftenOnLab`
- `New-OftenOnLab`
- `Stop-OftenOnLab`
- `Start-OftenOnLab`
- `Remove-OftenOnLab`

## Demo

![Building the OftenOn Lab][11]

`Set-OftenOnLab -ConfigurationName X` lets you switch to different migration scenarios.
- `Default` is the original OftenOn, a five-node no-storage multi-subnet WSFC cluster with WS2012 and SQL2012.
- `CrossClusterMigration` is a WS2012 and SQL2012 cluster, with a second WS2016 and SQL2012 cluster. You can use this to practice
SQL2012 cross-cluster migrations.
- `Upgrade` is a WS2012 and SQL2012 cluster, with a second WS2016 and SQL2017 cluster. You can use this to practice upgrading
SQL2012 to SQL2017 and then setting up a Distributed Availability Group to migrate from one cluster to another.
- `DAG` is a WS2012 and SQL2017 cluster, with a second WS2016 and SQL2017 cluster. You can use this to practice just setting up
a Distributed Availability Group and migrating from one cluster to another.

It also takes an optional `-ModulePath`. You can point this to a folder and the contents will be copied into C:\Program Files\WindowsPowerShell on the workstation machine. This is a quick way to get your own code into there.

Then execute `New-OftenOnLab`.

It takes time on the first run to download about 10GB of Evaluation ISOs from Microsoft. After this though each time the lab is
destroyed and competely recreated takes 45-60 minutes. You'll know it's done when all the VMs go to the login screen.

![Show the OftenOn VMs][12]

Then remote in yourself! Remote Desktop from your machine should work to the short computer names (e.g. CHDC01) via IPv6 but it may
not work on some configurations.

Username OFTENON\LocalAdministrator
Password Local2019!

## VM naming conventions

The VMs built are as follows:

- CHDC01 the domain controller in the Chicago subnet that also does routing for other subnets.
- CHWK01 a workstation in Chicago with .NET 4.7.2 and SSMS 18.2.
- SEC1N1 SEC1N2 SEC1N3 are three nodes of the C1 cluster in the Seattle subnet.
- DAC1N1 DAC1N2 are two nodes of the C1 cluster in the Dallas subnet.

(and for the non-default scenarios)
- SEC2N1 SEC2N2 SEC2N3 are three nodes of the C2 cluster in the Seattle subnet.
- DAC2N1 DAC2N2 are two nodes of the C2 cluster in the Dallas subnet.

The C1 cluster has an Availability Group (AG1), a listener (AG1L), and 4 databases (AG1DBx). It's all sync'd to Dallas and ready
to failover for your testing! C2 cluster usually doesn't have an Availability Group because setting that up is part of doing your
migration. Where it is required, I recomment set it up like so:

- Availability Group AG2
- Listener AG2L
- IP Addresses 10.0.1.131/255.255.255.0 and 10.0.2.131/255.255.255.0

And if you use a Distributed Availability Group:

- DAG1 (covering AG1 and AG2)

## Other notes

All VMs have WMF 5.1 installed. There's a couple resources (like SQL patches and the .NET 4.7.2 updater) on \\CHDC01\Resources.
Those aren't installed everywhere because that's part of the fun of the lab - exploring various issues with and without them.
It's also loaded with all my other PowerShell modules.

![Show the OftenOn AG][13]

[1]: Images/oftenon.ai.svg
[3]: CHANGELOG.md

[11]: Images/oftenon1.gif
[12]: Images/oftenon2.gif
[13]: Images/oftenon3.gif


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.16] - 2020-02-19
### Fixed
- Default2017 configuration fixed.

## [1.1.15] - 2020-02-19
### Changed
- SSMS 18.4, SQL 2017 CU19, new Default2017 configuration (one SQL 2017 set
  on Windows 2016).
- DSC resource updates.

## [1.1.14] - 2019-09-30
### Changed
- SSMS 18.3

## [1.1.13] - 2019-09-20
### Changed
- SSMS to 18.2 and SQL 2017 to CU16.

## [1.1.12] - 2019-09-20
### Changed
- Module version updates.

## [1.1.11] - 2019-09-20
### Changed
- Module version updates.

## [1.1.10] - 2019-09-20
### Changed
- Module version updates.

## [1.1.9] - 2019-09-20
### Changed
- DSC module version updates.

### Added
- Set-OftenOnLab now has a -Subnet parameter that lets you switch between the
  default subnet of 10.0 to 192.168 or 172.16.

## [1.1.8] - 2019-09-04
### Fixed
- Re-remove content from OftenOn.psm1 as I forgot it's in Test-OftenOnLab.

## [1.1.7] - 2019-08-30
### Fixed
- Its own update. Sigh.

## [1.1.6] - 2019-08-30
### Fixed
- DSC and other module version updates.

## [1.1.5] - 2019-08-30
### Fixed
- Re-added missing header to OftenOn.psm1 which does version checksk
- DSC module version updates.

## [1.1.4] - 2019-06-20
### Fixed
- ISO checksum for SQL 2017.

## [1.1.3] - 2019-06-19
### Fixed
- Re-address the NlaSvc "this network is Public" issue with a startup script.
- Re-address the application of DSC configuration after reboot with another
  startup script.

## [1.1.2] - 2019-06-18
### Changed
- Updated dependent module numbers.

## [1.1.1] - 2019-06-17
### Changed
- Updated dependent module numbers.

## [1.1.0] - 2019-06-17
### Added
- Set-OftenOnLab can be used to alter what configuration New-OftenOnLab will
  create. Options are:
    Default (SQL 2012 on Windows 2012)
    CrossClusterMigration (2x clusters of SQL 2012 on Windows 2012)
    Upgrade (SQL 2012 on Windows 2012 and SQL 2017 on Windows 2016)
    DAG (SQL 2017 on Windows 2012 and SQL 2017 on Windows 2017)
- Windows 2016 support.

### Changed
- SQL patches are applied by default because a lot of the migrations require
  them.
- CHDC01 and CHWK01 are now Windows 2016. This makes life easier as we are
  only really interested in migration scenarios where SQL is on Windows 2012.
- The domain LAB is now called OFTENON. This bears no relation to any .com of
  the same name.
- SSMS 18.1.0 is installed by default. The previous two are still available.
- 4 databases are added to the Availability Group by default.
- Added version numbers to some non-DSC modules which are copied in. This is
  because Lability is caching the old versions.

### Fixed
- Internet access for machines inside the domain. This is critical for time sync
  which is critical for Windows 2016 support.
  If it doesn't work for you and you're on an old Windows 10 version like 1607,
  modify your Default Switch in Hyper-V manager to be an External switch. On
  newer Windows 10 it should just work without this.

## [1.0.21] - 2019-04-18
### Fixed
- Revert broken version.

## [1.0.19] - 2019-04-15
### Removed
- HackSql module is no longer installed by default in VMs as it can be flagged
  as an unwanted program by AV.

## [1.0.18] - 2019-04-04
### Changed
- Updated for latest DSC Resource Kit release.

## [1.0.17] - 2019-04-04
### Changed
- Updated for latest DSC Resource Kit release.

## [1.0.16] - 2019-04-03
### Changed
- VM startup memory reduced to 1GB so that builds are more consistent on a
  16GB laptop. This means BootOrder and BootDelay was no longer needed.

### Fixed
- The new bootstrap scheduled task removes itself once DSC reports that it is
  complete. This prevents any of your own customisations from possibly being
  overwritten on a later reboot.
- WAN adapter.

### Broken
- Desired state is never reached on CHDC01.

 VERBOSE: [CHDC01]:
 [[DnsServerAddress]SetDnsServerAddressDALLAS_HB::[ooNetwork]RenameNetwork]
 Test-TargetResource: DNS server addresses are not correct.
 Expected "", actual "127.0.0.1".


## [1.0.15] - 2019-03-09
### Changed
- Internet Explorer Enhanced Security Configuration is disabled by default.
- Passwords end with 2019 instead of 2018.
- All nodes now have internet access that go through the router/NAT on the
  DC. This is set up on the DC *after* the domain has been created. Doing
  this before the domain has been created stops it from being created and
  stops nodes from joining.
- There's a scheduled task which re-applies the Bootstrap DSC configuration
  on every boot. This is to make the initial configuration go faster and
  also results in better capturing of errors to the text logs.

### Notes
- After ISOs and others have been downloaded once, this now builds the lab
  in 56 minutes on my laptop. Here's the timing.

  00 minutes - MOFs compile and VMs get created.
  05 minutes - VMs start. CHDC01 domain creation starts.
  20 minutes - CHDC01 reboots followed by all other nodes as the pick up
               the new domain. SQL installs start.
  55 minutes - Finished.

### Added
- Stop-OftenOnLab has a new -TurnOff parameter which is faster.
- Remove-OftenOnLab now uses Stop-OftenLab -TurnOff and so runs faster.

### Broken
- The DC WAN adapter is now broken and has no internet access. I don't know why.

## [1.0.14] - 2019-02-27
### Fixed
- LCM DebugMode set to ForceModuleImport. This will massively speed up setup
  by not caching DSC responses from resources, which slows down joining the
  domain and setting up the cluster.

### Changed
- DSC resource module version number updates.
- Don't quit when the DSC modules are wrong. This is so you can still use
  Remove-OftenOnLab when things are out of date.

## [1.0.13] - 2019-02-13
### Changed
- Updated required version of modules, and give more errors upon load if the

psakefile.ps1
Task default -depends Compile

Set-StrictMode -Version Latest
$ErrorActionPreference = "Stop"

#region Network translation functions
function ConvertFrom-CIDR {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [ValidateScript( { $_ -match "(.*)\/(\d+)" })]
        [string] $IPAddress
    )

    Write-Verbose ($IPAddress -match "(.*)\/(\d+)")
    $ip = [IPAddress] $Matches[1]
    $suffix = [int] $Matches[2]
    $mask = ("1" * $suffix) + ("0" * (32 - $suffix))
    $mask = [IPAddress] ([Convert]::ToUInt64($mask, 2))

    Write-Verbose "IP $ip CIDR $suffix MASK $mask"

    @{
        IPAddress  = $ip
        CIDR       = $IPAddress
        CIDRSuffix = $suffix
        NetworkID  = ([IPAddress] ($ip.Address -band $mask.Address)).IPAddressToString
        SubnetMask = $mask.IPAddressToString
    }
}

function ConvertTo-CIDR {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [string] $IPAddress,
        [Parameter(Mandatory)]
        [string] $SubnetMask
    )

    $ip = [IPAddress] $IPAddress
    $mask = [IPAddress] $SubnetMask
    Write-Verbose ($mask.IPAddressToString -match '(.*)\.(.*)\.(.*)\.(.*)')
    $suffix = ""
    $Matches[1..4] | ForEach-Object {
        $suffix += [Convert]::ToString([int] $_, 2) + ("0" * (8 - [Convert]::ToString([int] $_, 2).Length))
    }
    $suffix = ($suffix -split "[^1]")[0].Length

    @{
        IPAddres   = $IPAddress
        CIDR       = "$IPAddress/$suffix"
        CIDRSuffix = $suffix
        NetworkID  = ([IPAddress] ($ip.Address -band $mask.Address)).IPAddressToString
        SubnetMask = $mask.IPAddressToString
    }
}
#endregion

#region Always-run, manipulate configuration data encryption information
$configurationData = Import-PowerShellDataFile -Path C:\Lability\Configurations\WS2012.psd1
$configurationData.AllNodes | Where-Object { $_.NodeName -eq '*' } | ForEach-Object {
    $PSItem.CertificateFile = $PSItem.CertificateFile.Replace('$env:ALLUSERSPROFILE', $env:ALLUSERSPROFILE)
    $certificate = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2
    $certificate.Import($PSItem.CertificateFile)
    $PSItem.Thumbprint = $certificate.Thumbprint
}
#endregion

#region Always-run, manipulate configuration data network information
foreach ($node in $configurationData.AllNodes) {
    <#
        Lability creates one NIC for each entry in the Lability_SwitchName array. We put these into a Network
        array instead so that we have more fine-grained control.

        We use that to populate Lability_SwitchName and Lability_MACAddress which assigns a random MAC. This
        is used later to match up with the network AdapterName and rename the NIC.

        This must be done because most versions of Windows assign default NIC names in an uncontrolled manner.
    #>
    if ($node.ContainsKey('Network')) {
        $switchName = @()
        $macAddress = @()

        foreach ($network in $node.Network) {
            $switchName += $network.SwitchName
            # It's important to limit what MAC are used otherwise you will get confusing errors during VM creation
            $macAddress += ('00', '03', (0..3 | ForEach-Object { '{0:x}{1:x}' -f (Get-Random -Minimum 0 -Maximum 15), (Get-Random -Minimum 0 -Maximum 15) }) | ForEach-Object { $_ }) -join ':'
        }

        $node.Lability_SwitchName = $switchName
        $node.Lability_MACAddress = $macAddress
    }

    <#
        Different xFailoverCluster resources need different combinations of IP, CIDR, and Subnet. This splits
        out all of the different variations for easy use during configuration.
    #>
    if ($node.ContainsKey('Role') -and $node.Role.ContainsKey('Cluster')) {
        $node.Role.Cluster.StaticAddress = ConvertFrom-CIDR $node.Role.Cluster.StaticAddress
        $node.Role.Cluster.StaticAddress.Name = "Cluster Network " + ($node.Network | Where-Object { (ConvertFrom-CIDR $_.IPAddress).NetworkID -eq $node.Role.Cluster.StaticAddress.NetworkID }).NetAdapterName + " (Client)"
        $node.Role.Cluster.IgnoreNetwork = ConvertFrom-CIDR $node.Role.Cluster.IgnoreNetwork
        $node.Role.Cluster.IgnoreNetwork.Name = "Cluster Network " + ($node.Network | Where-Object { (ConvertFrom-CIDR $_.IPAddress).NetworkID -eq $node.Role.Cluster.IgnoreNetwork.NetworkID }).NetAdapterName + " (Heartbeat)"
    }
}
#endregion

#region Apply a local RDP fix so that it can connect to unpatched RDP servers such as will be in our VMs.
Task FixRDP {
    if (!(Test-Path HKLM:\Software\Microsoft\Windows\CurrentVersion\Policies\System\CredSSP)) {
        New-Item HKLM:\Software\Microsoft\Windows\CurrentVersion\Policies\System\CredSSP
    }
    if (!(Test-Path HKLM:\Software\Microsoft\Windows\CurrentVersion\Policies\System\CredSSP\Parameters)) {
        New-Item HKLM:\Software\Microsoft\Windows\CurrentVersion\Policies\System\CredSSP\Parameters
    }
    Set-ItemProperty -Path 'HKLM:\Software\Microsoft\Windows\CurrentVersion\Policies\System\CredSSP\Parameters' -Name 'AllowEncryptionOracle' -Value 2 -Type DWord
}
#endregion

#region Just compile the MOF resources
Task Compile {
    . .\WS2012.ps1
    WS2012 -ConfigurationData $configurationData -OutputPath C:\Lability\Configurations
}
#endregion

#region Compile the MOF resources and configure the lab VMs
Task Build -depends Compile {
    $administrator = New-Object System.Management.Automation.PSCredential('Administrator', ('Admin2018!' | ConvertTo-SecureString -AsPlainText -Force))
    Start-LabConfiguration -ConfigurationData $configurationData -IgnorePendingReboot -Credential $administrator -NoSnapshot
}
#endregion

#region Start the lab VMs
Task Start {
    "Allow at least 15 minutes for the first boot to finish Domain Controller setup"
    Start-Lab -ConfigurationData $configurationData
}
#endregion

#region Remove existing VMs, then compile the MOF resources, configure the lab VMs, and start the lab VMs
Task BuildAll -Depends Clean, Build, Start {
}
#endregion

#region Stop the lab VMs
Task Stop {
    Stop-Lab -ConfigurationData $configurationData -ErrorAction:SilentlyContinue
}
#endregion

#region Remove existing VMs
Task Clean -depends Stop {
    Remove-LabConfiguration -ConfigurationData C:\Lability\Configurations\WS2012.psd1 -ErrorAction:SilentlyContinue -Confirm:$false
    Remove-Item C:\Lability\VMVirtualHardDisks\*
}
#endregion

#region Remove existing VMs and also master hard disks so the server OS is rebuilt
Task CleanAll -depends Clean {
    Remove-Item C:\Lability\MasterVirtualHardDisks\*
}
#endregion

Task Composite {
    $configuration = Read-Host "Configuration name"
    New-Item -ItemType Directory C:\Git\OftenOn\DSCResources\$configuration
    New-ModuleManifest "C:\Git\OftenOn\DSCResources\$configuration\$($configuration).psd1" -RootModule ".\$($configuration).schema.psm1" -Author "Cody Konior" -CompanyName ""
    $content = "Configuration $configuration {
}
"
    [System.IO.File]::WriteAllLines("C:\Git\OftenOn\DSCResources\$configuration\$($configuration).schema.psm1", $content)
    code "C:\Git\OftenOn\DSCResources\$configuration\$($configuration).schema.psm1"
}

<#

Add node to cluster can have an error which triggers a 15 minute wait.

TODO
    Set which node is in sync/auto to start with
    Stop the SQL Server AG install unless all nodes are joined
    Set up WSUS

    # Give two options, either IP Forwarding (no WAN) or
    Install-RemoteAccess -VpnType Vpn
    &netsh routing ip nat install
    $ExternalInterface = 'WAN'
    &netsh routing ip nat add interface $ExternalInterface
    &netsh routing ip nat set interface $ExternalInterface mode=full
    ^-- This seems to be enough even with IP forwarding disabled, also gives NAT
    But then you need to enable recursion, and add a forwarder to 1.1.1.1
    # $InternalInterface1 = 'LAN1'
    # $InternalInterface2 = 'LAN2'
    # cmd.exe /c 'netsh routing ip nat add interface $InternalInterface1'
    # cmd.exe /c 'netsh routing ip nat add interface $InternalInterface2'
    # xDnsServerForwarder "WAN Forwarder" {
    #     IsSingleInstance = "Yes";
    #     IPAddresses = "1.1.1.1";
    #     # DependsOn = '[WindowsFeature][DNS]';
    # }

    Create MSA accounts
        Add SecurityPolicyDsc permissions
        Create RDCMan
        Try adding a WAN card

#>
